i18n: Add `resetLocaleData` method

yuliyan opened the pull request:
## Description
Currently, it's not possible to reset the locale data of a `@wordpress/i18n` instance. The `setLocaleData` method that is supposed to be used to modify the data, would only merge the new and the existing data and wouldn't allow to reset the data that has already been loaded to the instance.

This PR adds new `resetLocaleData` method to the @wordpress/i18n instance that resets all current Tannin instance locale data and sets the new locale data for specified the domain.

## How has this been tested?
* Existing unit tests pass.
* Added new tests that should cover `resetLocaleData` funcitonality.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

yuliyan committed:
i18n: Add `setLocaleData` merge data flag

swissspidy commented:
The only question I have here right now is whether this should be its own separate method instead of such a flag, i.e. `resetlocaleData`/`removeLocaleData`/`clearLocaleData`.

yuliyan commented:
> The only question I have here right now is whether this should be its own separate method instead of such a flag, i.e. `resetlocaleData`/`removeLocaleData`/`clearLocaleData`.

I thought if implementing it as a separate method initially, but after @jsnajdr suggested it as an option [in another discussion](https://github.com/Automattic/wp-calypso/pull/51406#discussion_r604877352) , I liked the boolean argument approach more. I don't really have a strong preference on which approach should be implemented, but my personal arguments for using the flag implementation against the separate method approach would be that resetting the locale data would usually be followed by a `setLocaleData` call anyway (e.g. switching to a different language use case). Regarding the code, I don't think there would much difference. Maybe it saves some code repetition, but it's not a significant difference, imo.

I'll be happy to hear other thoughts on the topic.

jsnajdr commented:
One thing that I don't like on the `setLocaleData( data, domain, reset )` signature is that the `domain` and `reset` parameters kind of get in each other's way. `domain` is optional and has a default, and when specifying `reset`, I have no choice but to specify it, if only as `undefined`.

From that point of view, `resetLocaleData( data, domain )` looks more symmetrical. It should reset the instance state completely. If there is state other than `this.tannin.data`, it should reset it, too. Making it equivalent to simply recreating the object with `createI18n( data, domain )`.

jsnajdr commented on packages/i18n/src/create-i18n.js:
```diff
@@ -158,9 +158,19 @@ export const createI18n = ( initialData, initialDomain, hooks ) => {
+		if ( ! shouldMerge ) {
+			// Remove the existing locale data for the specified domain.
+			delete tannin.data[ domain ];
+		}
```
This is one place where `setLocaleData( data, domain, false )` is not equivalent to recreating the `I18n` instance. Data for other domains would remain in place.

Given that the use case that led to this new API is changing the language of the instance, it would make more sense to remove other domains' data, too.

yuliyan committed:
Revert "i18n: Add `setLocaleData` merge data flag"

This reverts commit 978685c15efd6342f990a1829bb70ae485208e23.

yuliyan committed:
i18n: Add `resetLocaleData` method

Adds new `resetLocaleData` method to the @wordpress/i18n instance that resets all current Tannin instance locale data and sets the new locale data for specified the domain.

yuliyan commented:
Thank you both for the feedback. I've reworked the code and updated the PR description. It now adds the new `resetLocaleData` method.

swissspidy commented on packages/i18n/src/create-i18n.js:
```diff
@@ -181,6 +191,14 @@ export const createI18n = ( initialData, initialDomain, hooks ) => {
 		notifyListeners();
 	};
 
+	/** @type {SetLocaleData} */
```
```suggestion
	/** @type {ResetLocaleData} */
```

yuliyan committed:
Use correct type

Co-authored-by: Pascal Birchler <pascal.birchler@gmail.com>

jsnajdr commented on packages/i18n/src/create-i18n.js:
```diff
@@ -181,6 +191,14 @@ export const createI18n = ( initialData, initialDomain, hooks ) => {
+	/** @type {ResetLocaleData} */
+	const resetLocaleData = ( data, domain ) => {
+		// Reset all current Tannin locale data .
+		tannin.data = {};
```
When resetting the `Tannin` instance, we should also set `tannin.pluralForms = {}`. It is a per-domain cache of plural form data derived from `this.data`. See how the cache is used here: https://github.com/aduth/tannin/blob/master/packages/tannin/index.js#L134-L168

yuliyan replied:
Thanks, I wasn't aware of this caching. Updated the code and added some tests.

yuliyan committed:
Reset plural forms function cache

jsnajdr reviewed (approved):
Looks ready to ship :+1: I apologize for being slow to respond 🐌

yuliyan commented:
> Looks ready to ship 👍 I apologize for being slow to respond 🐌

@jsnajdr, I'm not authorized to do the merge, so it should be done by you or @swissspidy.
